### PR TITLE
Trampolines for large binary expressions

### LINKED ISSUE #35633: Stack overflow in bindBinaryExpressionFlow
From telemetry, ~~seems to be new in 3.6~~ (not sure when it was introduced, but we started collecting data around the 3.6 timeframe). It's a high hitter in VS.

User report: https://developercommunity.visualstudio.com/content/problem/850916/javascript-error-on-intellisense-initation.html

```
Error processing request. Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
at Object.getAssignmentDeclarationKind (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:11540:42)
at bindWorker (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30970:42)
at bind (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30867:13)
at visitNode (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17584:24)
at Object.forEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17804:24)
at bindEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29481:16)
at bindBinaryExpressionFlow (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30180:17)
at bindChildrenWorker (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29542:21)
at bindChildren (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29451:17)
at bind (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30878:21)
at visitNode (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17584:24)
at Object.forEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17804:24)
at bindEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29481:16)
at bindBinaryExpressionFlow (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30180:17)
at bindChildrenWorker (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29542:21)
at bindChildren (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29451:17)
at bind (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30878:21)
at visitNode (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17584:24)
at Object.forEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17804:24)
at bindEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29481:16)
at bindBinaryExpressionFlow (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30180:17)
at bindChildrenWorker (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29542:21)
at bindChildren (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:29451:17)
at bind (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:30878:21)
at visitNode (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17584:24)
at Object.forEachChild (C:\Program Files (x86)\Microsoft SDKs\TypeScript\3.7\tsserver.js:17804:24)
....
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #35633

This introduces trampolines into the binder, checker, and emitter. The binder trampoline is complete - every `bind` call returns to the root of the binder stack to be performed. The checker trampoline is partial - all immediately nested binary expressions are included within a single stack frame of `checkBinaryExpression`. Control flow is much less likely to need a trampoline (and does not in either given example), as the control flow graph is much simplified compared to the AST. However it could be added, if there is interest in doing so. The emitter trampoline is both partial and conditional; our printer handler API makes it awkward to try to trampoline through it, so to avoid breaking it, we only perform an emit trampoline for binary expressions when a node does not need notifications, is not substituted, does not need comment emit, and does not need sourcemap emit. This is quite limiting, but the primary concern here was typechecking and language server operations, so I figure having _a_ way to emit a file like those provided (with `removeComments: true`) without crashing might be better than always crashing during emit on them, on the offchance someone includes something like them in a project they're actually emitting. Do note though that we do still just.... run out of stack space, should those conditions not be met. If we're ever comfortable breaking our printer/transformer API, making it trampoline-able (either by return convention, similar to the trampoline in the emitter, or by generator construction, as in the binder) would be something we should do.

In addition, a few small other bits got trampolines.  In the parser, `fixupParentPointers` became a pretty trivial iterative algorithm, rather than the recursive one (this is only used during testing, so doesn't appear in the production stacktrace, but is required nonetheless). In the test harness, `visitNode` in the type/symbol writer now just utilizes a straight iterator of every AST node (which is built utilizing a `visitEachChild` generator-based variant produced to more easily implement implement the binder trampoline) without relying on any recursion.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot perf test this - I think my trampoline in the emitter is low/no overhead (and some of the smaller changes to enable it may actually lower stack/memory usage in normal scenarios), and the checker one isn't on a particularly hot path, so I don't expect much, if any, changes to check or emit time (and need to validate that). However the restructuring in the binder likely has performance implications; I am unsure of how much, so definitely need to check it out.

**Comment by weswigham** [comment]
@typescript-bot perf test this

**Comment by weswigham** [comment]
@typescript-bot perf test this now with 100% more runtime provided generators (do note: I'm still manually forwarding the generators in the binder because for some reason [that's faster than using `yield*`](https://jsperf.com/yield-vs-yield-in-for-loop)).

**Comment by weswigham** [comment]
Well, +130% is still horrible, but nowhere near as bad. Guess I'm really going to have to write trampolines the hard way, like I did in the emitter :(

**Comment by weswigham** [comment]
@typescript-bot perf test this

**Comment by weswigham** [comment]
There we go, much more reasonable numbers - variations between -2% to +5% on very small numbers. Much better. @rbuckton do you have opinions on this PR in this state?

**Comment by weswigham** [comment]
@typescript-bot perf test this again - now with less (unused) code in the parser (which maybe accounts for the parser time changes).

**Comment by weswigham** [comment]
@typescript-bot perf test this

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -1446,27 +1446,114 @@ namespace ts {
...
+                            res[4] = -1;
+                        }
+                        else {
+                            const savedSubtreeTransformFlags = subtreeTransformFlags;
+                            subtreeTransformFlags = 0;
+                            res[4] = savedSubtreeTransformFlags;
+                        }
+
+
+                        res[2] = saveInStrictMode;
```
Comment (rbuckton) [inline src/compiler/binder.ts]: I'd recommend either a const enum for field names in the tuple, or helper functions to get/set (if you are using parallel arrays), i.e.:

```ts
interface BindStack {
  frame: number;
  exprStack: BinaryExpression[];
  stateStack: ...;
  inStrictModeStack: (boolean | undefined)[];
  ...
}
function getInStrictMode(stack: BindStack) {
  return stack.inStrictModeStack[stack.frame];
}
function setInStrictMode(stack: BindStack, value?: boolean) {
  stack.inStrictModeStack[stack.frame] = value;
}

...

setInStrictMode(stack, saveInStrictMode);
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -27112,11 +27112,59 @@ namespace ts {
         }
 
         function checkBinaryExpression(node: BinaryExpression, checkMode?: CheckMode) {
-            if (isInJSFile(node) && getAssignedExpandoInitializer(node)) {
-                return checkExpression(node.right, checkMode);
+            const work: [BinaryExpression, number, Type | undefined][] = [[node, 0, undefined]];
```
Comment (rbuckton) [inline src/compiler/checker.ts]: As with the binder, I'd recommend parallel stacks.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -2425,19 +2427,79 @@ namespace ts {
...
+        /**
+         * emitBinaryExpression includes an embedded work stack to attempt to handle as many nested binary expressions
+         * as possible without creating any additional stack frames. This can only be done when the emit pipeline does
+         * not require notification/substitution/comment/sourcemap decorations.
+         */
         function emitBinaryExpression(node: BinaryExpression) {
-            const isCommaOperator = node.operatorToken.kind !== SyntaxKind.CommaToken;
-            const indentBeforeOperator = needsIndentation(node, node.left, node.operatorToken);
-            const indentAfterOperator = needsIndentation(node, node.operatorToken, node.right);
+            const work: [BinaryExpression, number][] = [[node, 0]];
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: I'd recommend parallel stacks.

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -2425,19 +2427,79 @@ namespace ts {
...
+            while (work.length) {
+                const res = work[work.length - 1];
+                let next: Expression | undefined;
+                node = res[0];
+                switch (res[1]) {
+                    case 0: {
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: I'd recommend using named states.

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -904,28 +904,37 @@ namespace ts {
             // overhead.  This functions allows us to set all the parents, without all the expense of
             // binding.
 
-            let parent: Node = rootNode;
-            forEachChild(rootNode, visitNode);
+            const stack: [Node, Node[]][] = [[rootNode, gatherChildren(rootNode)]];
```
Comment (rbuckton) [inline src/compiler/parser.ts]: I think you can do this with just a `const stack: Node[] = [rootNode]`. You can move the logic for checking/setting parents into this function and just push the children that still need to be visited onto the stack.

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -904,28 +904,37 @@ namespace ts {
...
-                    parent = n;
-                    forEachChild(n, visitNode);
-                    if (hasJSDocNodes(n)) {
-                        for (const jsDoc of n.jsDoc!) {
-                            jsDoc.parent = n;
-                            parent = jsDoc;
-                            forEachChild(jsDoc, visitNode);
+            function gatherChildren(node: Node) {
+                const children: Node[] = [];
+                forEachChild(node, n => void children.unshift(n)); // By using a stack above and `unshift` here, we emulate a depth-first preorder traversal
```
Comment (rbuckton) [inline src/compiler/parser.ts]:
```suggestion
                forEachChild(node, n => { children.unshift(n); }); // By using a stack above and `unshift` here, we emulate a depth-first preorder traversal
```

File: src/harness/typeWriter.ts [hunk 7]
```diff
@@ -21,6 +21,24 @@ namespace Harness {
...
+    function* forEachASTNode(node: ts.Node) {
+        const work = [[node]];
+        while (work.length) {
+            const res = work[work.length - 1];
+            if (res.length > 0) {
+                const elem = res.shift()!;
+                yield elem;
+
+                const resChildren: ts.Node[] = [];
+                ts.forEachChild(elem, c => void resChildren.push(c));
```
Comment (rbuckton) [inline src/harness/typeWriter.ts]:
```suggestion
                ts.forEachChild(elem, c => { resChildren.push(c); });
```

File: src/harness/typeWriter.ts [hunk 8]
```diff
@@ -21,6 +21,24 @@ namespace Harness {
         type?: string;
     }
 
+    function* forEachASTNode(node: ts.Node) {
+        const work = [[node]];
```
Comment (rbuckton) [inline src/harness/typeWriter.ts]: You might be able to do this with just an array of `Node`.
Just push the children onto `work` in reverse order to maintain pre-order traversal in the DFS

**Comment by weswigham** [comment]
@typescript-bot perf test this

**Comment by weswigham** [comment]
@rbuckton perf looks good and your changes are in - anything else to add?

File: src/compiler/binder.ts [hunk 9]
```diff
@@ -1445,28 +1445,156 @@ namespace ts {
...
+                inStrictMode: [undefined],
+                parent: [undefined],
+                subtreeFlags: [undefined]
+            };
+            let stackIndex = 0;
+            while (stackIndex >= 0) {
+                node = workStacks.expr[stackIndex];
+                switch (workStacks.state[stackIndex]) {
+                    case BindBinaryExpressionFlowState.BindThenBindChildren: {
+                        // The -1 state is used only when recuring, to emulate the work that `bind` does before
```
Comment (rbuckton) [inline src/compiler/binder.ts]: nit: "-1 state" doesn't make sense here due to the named states.

File: src/compiler/binder.ts [hunk 10]
```diff
@@ -1445,28 +1445,156 @@ namespace ts {
...
+                        // reaching `bindChildren`. A normal call to `bindBinaryExpressionFlow` will already have done this work.
+                        node.parent = parent;
+                        const saveInStrictMode = inStrictMode;
+                        bindWorker(node);
+                        const saveParent = parent;
+                        parent = node;
+
+                        let subtreeFlagsState: number | undefined;
+                        // While this next part does the work of `bindChildren` before it descends into `bindChildrenWorker`
+                        // and uses `res[4]` to queue up the work that needs to be done once the node is bound.
```
Comment (rbuckton) [inline src/compiler/binder.ts]: nit: the term "`res[4]`" doesn't make sense with the change to parallel stacks.

File: src/compiler/transformer.ts [hunk 11]
```diff
@@ -288,12 +289,7 @@ namespace ts {
         function emitNodeWithNotification(hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) {
             Debug.assert(state < TransformationState.Disposed, "Cannot invoke TransformationResult callbacks after the result is disposed.");
             if (node) {
-                if (isEmitNotificationEnabled(node)) {
-                    onEmitNode(hint, node, emitCallback);
-                }
-                else {
-                    emitCallback(hint, node);
-                }
+                onEmitNode(hint, node, emitCallback);
```
Comment (rbuckton) [inline src/compiler/transformer.ts]: There is a negative aspect of moving this to `TransformationResult` and `PrintHandlers`, in that any existing API consumers *aren't* doing this, so we could potentially break them by removing this check. We should err on the side of caution and leave this check here (and merely use the new `PrintHandlers` hook as an optimization).

**Comment by weswigham** [comment]
@rbuckton added the check back in (and fixed the comments)

**Comment by minestarks** [comment]
Using `3.8.0-dev.20200208`, I'm still getting a stack overflow for the large binary expression in https://github.com/openlayers/ol2/blob/master/tests/speed/wmscaps.js  . Was this change not supposed to address that? 

```
    Maximum call stack size exceeded

    RangeError: Maximum call stack size exceeded
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98711:30)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
        at visitNode (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18318:24)
        at Object.forEachChild (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:18538:24)
        at walk (e:\temp\ol2\node_modules\typescript\lib\tsserver.js:98794:24)
```

**Comment by weswigham** [comment]
It only fully removes the crash with the caveats that either `noEmit: true` is set or `removeComments: true` is set, due to some limitations in our API layering for emit. It's also possible there's another transform that's enabled that also hasn't been patched with a trampoline? (Afaik, I just targeted es3 with the unittest)

**Comment by minestarks** [comment]
@weswigham hmmmmmm


```
E:\temp\ol2>node_modules\.bin\tsc.cmd -v
Version 3.8.0-dev.20200208

E:\temp\ol2>node_modules\.bin\tsc.cmd --allowJs --noEmit tests\speed\wmscaps.js
E:\temp\ol2\node_modules\typescript\lib\tsc.js:81132
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14658:23)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14865:24)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14865:24)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14865:24)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
```

**Comment by minestarks** [comment]
Here's a modified repro so node prints out the full stack:

```
>node --stack-trace-limit=1000 --stack-size=100  node_modules\typescript\bin\tsc   --allowJs --noEmit tests\speed\wmscaps.js

E:\temp\ol2\node_modules\typescript\lib\tsc.js:81132
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81199:30)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14865:24)

... yada yada yada ...

    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14865:24)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14739:21)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at walkArray (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81335:25)
    at visitNodes (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14664:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14894:24)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at visitNode (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14659:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14892:21)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at walkArray (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81335:25)
    at visitNodes (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14664:24)
    at Object.forEachChild (E:\temp\ol2\node_modules\typescript\lib\tsc.js:14887:24)
    at walk (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81278:24)
    at E:\temp\ol2\node_modules\typescript\lib\tsc.js:81197:17
    at runWithCancellationToken (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81125:24)
    at getJSSyntacticDiagnosticsForFile (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81194:20)
    at getSyntacticDiagnosticsForFile (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81117:65)
    at E:\temp\ol2\node_modules\typescript\lib\tsc.js:81073:24
    at Object.flatMap (E:\temp\ol2\node_modules\typescript\lib\tsc.js:388:25)
    at getDiagnosticsHelper (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81069:56)
    at Object.getSyntacticDiagnostics (E:\temp\ol2\node_modules\typescript\lib\tsc.js:81077:20)
    at emitFilesAndReportErrors (E:\temp\ol2\node_modules\typescript\lib\tsc.js:84730:45)
    at Object.emitFilesAndReportErrorsAndGetExitStatus (E:\temp\ol2\node_modules\typescript\lib\tsc.js:84765:18)
    at performCompilation (E:\temp\ol2\node_modules\typescript\lib\tsc.js:87119:29)
    at executeCommandLineWorker (E:\temp\ol2\node_modules\typescript\lib\tsc.js:87022:17)
    at Object.executeCommandLine (E:\temp\ol2\node_modules\typescript\lib\tsc.js:87049:20)
    at Object.<anonymous> (E:\temp\ol2\node_modules\typescript\lib\tsc.js:87283:4)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (E:\temp\ol2\node_modules\typescript\bin\tsc:2:1)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:282:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:743:3)
```

**Comment by minestarks** [comment]
@weswigham argh, only repros for JS files because of `getJSSyntacticDiagnosticsForFile` 🤦‍♀

**Comment by weswigham** [comment]
welp, that's a new callstack to me, XD

The `walk` function in `getJSSyntacticDiagnosticsForFile` needs to be rewritten to be a loop, rather than recursive calls~

**Pull request merged**